retrievalmarket/impl/providerstates/provider_states.go

```diff
@@ -39,40 +41,125 @@ func UnsealData(ctx fsm.Context, environment ProviderDealEnvironment, deal rm.Pr
-	}
+
 	if deal.ChannelID != nil {
 		log.Debugf("resuming data transfer for deal %d", deal.ID)
-		err = environment.ResumeDataTransfer(ctx.Context(), *deal.ChannelID)
+		err := environment.ResumeDataTransfer(ctx.Context(), *deal.ChannelID)
 		if err != nil {
 			return ctx.Trigger(rm.ProviderEventDataTransferError, err)
 		}
 	}
 	return nil
 }
 
-// CancelDeal clears a deal that went wrong for an unknown reason
-func CancelDeal(ctx fsm.Context, environment ProviderDealEnvironment, deal rm.ProviderDealState) error {
-	// Read next response (or fail)
-	err := environment.UntrackTransfer(deal)
+// UpdateFunding saves payments as needed until a transfer can resume
+func UpdateFunding(ctx fsm.Context, environment ProviderDealEnvironment, deal rm.ProviderDealState) error {
+	log.Debugf("handling new event while in ongoing state of transfer %d", deal.ID)
+	if deal.ChannelID == nil {
+		return nil
+	}
+	channelState, err := environment.ChannelState(ctx.Context(), *deal.ChannelID)
 	if err != nil {
 		return ctx.Trigger(rm.ProviderEventDataTransferError, err)
 	}
-	err = environment.DeleteStore(deal.ID)
+	result := updateFunding(ctx, environment, deal, channelState)
+	return environment.UpdateValidationStatus(ctx.Context(), *deal.ChannelID, result)
+}
+
+func updateFunding(ctx fsm.Context,
+	environment ProviderDealEnvironment,
+	deal rm.ProviderDealState,
+	channelState datatransfer.ChannelState) datatransfer.ValidationResult {
+	received, err := processLastVoucher(ctx, environment, channelState)
+	if err != nil {
+		return errorDealResponse(deal.Identifier(), err)
+	}
+
+	totalPaid := big.Add(deal.FundsReceived, received)
+
+	// check if all payments are received to continue the deal, or send updated required payment
+	owed := deal.Params.OutstandingBalance(totalPaid, channelState.Queued(), channelState.Status().InFinalization())
+	log.Debugf("provider: owed %d, total received %d = received so far %d + newly received %d, unseal price %d, price per byte %d",
+		owed, totalPaid, deal.FundsReceived, received, deal.UnsealPrice, deal.PricePerByte)
+
+	var voucherResult datatransfer.VoucherResult
+	if owed.GreaterThan(big.Zero()) {
+		if received.GreaterThan(big.Zero()) {
+			log.Debugf("provider: owed %d: sending partial payment request", owed)
+			_ = ctx.Trigger(rm.ProviderEventPartialPaymentReceived, received)
+		}
+		if received.GreaterThan(big.Zero()) || deal.Status != rm.DealStatusFundsNeededUnseal {
+			voucherResult = &rm.DealResponse{
+				ID:          deal.ID,
+				Status:      deal.Status,
+				PaymentOwed: owed,
+			}
+		}
+	} else {
+		_ = ctx.Trigger(rm.ProviderEventPaymentReceived, received)
+		if deal.Status == rm.DealStatusFundsNeededLastPayment {
+			log.Debugf("provider: funds needed: last payment")
+			voucherResult = &rm.DealResponse{
+				ID:     deal.ID,
+				Status: rm.DealStatusCompleted,
+			}
+		}
+	}
+	return datatransfer.ValidationResult{
+		Accepted:             true,
+		VoucherResult:        voucherResult,
```

Comment:
it's ok that this could be an empty voucherResult?

Reply:
yes. Nil is an acceptable voucher result